### PR TITLE
[GPU] Add a pass converting specific custom calls to custom fusions.

### PR DESCRIPTION
**Pull request by sergachev**
This will let JAX users run chosen computations as cuDNN kernels.

### CODE REVIEW

File: xla/service/gpu/custom_call_to_fusion_converter.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+#include "absl/strings/string_view.h"
+#include "xla/hlo/ir/hlo_module.h"
+#include "xla/service/hlo_pass_interface.h"
+
+namespace xla {
+namespace gpu {
+
+// Converts custom calls with specific backend configs (currently cuDNN) to
+// fusions with the same backend config.
+class CustomCallToFusionConverter : public HloModulePass {
```
Comment (jreiffers) [inline xla/service/gpu/custom_call_to_fusion_converter.h:30]: Can we give this a more specific name? Do you have specific plans to add more custom call types here?

File: xla/service/gpu/custom_call_to_fusion_converter.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+#include "absl/container/flat_hash_set.h"
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+#include "xla/hlo/ir/hlo_module.h"
+#include "xla/service/hlo_pass_interface.h"
+
+namespace xla {
+namespace gpu {
+
+// Converts custom calls with specific backend configs (currently cuDNN) to
```
Comment (jreiffers) [inline xla/service/gpu/custom_call_to_fusion_converter.h:28]: Can you add a sentence or two to describe the purpose of this transformation?

File: xla/service/gpu/custom_call_to_fusion_converter.h (hunk 1 above)
Comment (sergachev) [reply xla/service/gpu/custom_call_to_fusion_converter.h]: Done. Will probably add more later, no specific plan yet.

File: xla/service/gpu/custom_call_to_fusion_converter.h (hunk 2 above)
Comment (sergachev) [reply xla/service/gpu/custom_call_to_fusion_converter.h]: done

Comment (jreiffers) [reply xla/service/gpu/custom_call_to_fusion_converter.h]: Thanks a lot!

**Review by jreiffers — APPROVED** [review]

**Comment by sergachev** [comment]
resolved a conflict

**Review by jreiffers — APPROVED** [review]

File: xla/service/gpu/cudnn_custom_call_converter.cc [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+#include "absl/container/flat_hash_set.h"
+#include "absl/strings/string_view.h"
+#include "xla/hlo/ir/dfs_hlo_visitor_with_default.h"
+#include "xla/hlo/ir/hlo_computation.h"
+#include "xla/hlo/ir/hlo_instruction.h"
+#include "xla/hlo/ir/hlo_module.h"
+#include "xla/service/gpu/backend_configs.pb.h"
+#include "xla/service/gpu/ir_emission_utils.h"
+#include "xla/util.h"
```
Comment (akuegel) [inline xla/service/gpu/cudnn_custom_call_converter.cc:26]: There is a missing dependency for this header.

Comment (sergachev) [reply xla/service/gpu/cudnn_custom_call_converter.cc]: removed it

**Review by jreiffers — APPROVED** [review]

**Comment by sergachev** [comment]
Hi @jreiffers @akuegel! Is anything still blocking here?

**Comment by jreiffers** [comment]
@sergachev Looks like there's a conflict with the BUILD file.

**Comment by sergachev** [comment]
GitHub does not show any conflict; I tried rebasing to head locally and also did not get any conflict.

**Comment by jreiffers** [comment]
Hmm, it seems the imported version doesn't have https://github.com/openxla/xla/pull/15441 yet. Unfortunately I'm not very familiar with the tooling.

**Comment by jreiffers** [comment]
And the diff in here seems to have the same problem. Can you try rebasing and pushing the result to this PR?

**Comment by sergachev** [comment]
done

**Review by jreiffers — APPROVED** [review]

**Review by akuegel — APPROVED** [review]

File: xla/service/gpu/gpu_compiler.cc [hunk 4]
```diff
@@ -114,6 +114,7 @@ limitations under the License.
 #include "xla/service/gpu/autotuning/custom_kernel_fusion_autotuner.h"
 #include "xla/service/gpu/compile_module_to_llvm_ir.h"
 #include "xla/service/gpu/conv_layout_normalization.h"
+#include "xla/service/gpu/cudnn_custom_call_converter.h"
```
Comment (akuegel) [inline xla/service/gpu/gpu_compiler.cc:117]: There was a BUILD dependency missing for this header included. I am fixing this now, and will also move the pass to the transforms subdirectory. We are in the process of moving all HLO passes from the gpu directory to the transforms subdirectory.

**Comment by akuegel** [comment]
@sergachev It seems the new test is failing in debug mode. Unfortunately I will have to revert your change (we have a policy now that forbids fix forwards), but maybe you can already start looking into it before the revert is merged?

**Comment by sergachev** [comment]
I see, XLA does not like that I moved a computation from a custom call to a fusion, it wants it to be copied. I have a fix.

Thank you for helping.

**Pull request closed without merging**